File: macos/BolsaTerminal/Core/Models/History.swift
import Foundation

struct OHLCVBar: Codable, Hashable {
    let time: Double
    let open: Double
    let high: Double
    let low: Double
    let close: Double
    let volume: Double
}

// Warmup periods (e.g. EMA200 needs 200 bars) produce NaN on the backend,
// which `JSON.stringify` serializes as `null` — these must stay optional
// or decoding throws the moment a shorter period is requested.
struct IndicatorSet: Codable, Hashable {
    let rsi: [Double?]
    let macdLine: [Double?]
    let macdSignal: [Double?]
    let macdHistogram: [Double?]
    let sma20: [Double?]
    let sma50: [Double?]
    let ema20: [Double?]
    let ema200: [Double?]
    let bbUpper: [Double?]
    let bbMiddle: [Double?]
    let bbLower: [Double?]
    let bbPercent: [Double?]
    let volumeMA20: [Double?]
}

enum SignalType: String, Codable, CaseIterable {
    case strongBuy = "STRONG_BUY"
    case buy = "BUY"
    case neutral = "NEUTRAL"
    case sell = "SELL"
    case strongSell = "STRONG_SELL"

    var label: String {
        switch self {
        case .strongBuy: return "Compra fuerte"
        case .buy: return "Compra"
        case .neutral: return "Neutral"
        case .sell: return "Venta"
        case .strongSell: return "Venta fuerte"
        }
    }
}

struct SignalComponent: Codable, Hashable {
    let name: String
    let value: Double?
    let signal: String
    let score: Double
}

struct Recommendation: Codable, Hashable {
    let signal: SignalType
    let score: Double
    let components: [SignalComponent]
    let currentRSI: Double?
    let currentMACD: Double?
    let currentBBPercent: Double?
    let priceVsEMA200: Double?
}

struct HistoryResponse: Codable, Hashable {
    let symbol: String
    let bars: [OHLCVBar]
    let indicators: IndicatorSet
    let recommendation: Recommendation
}

File: android/core/model/src/main/java/com/bolsaterminal/core/model/History.kt
package com.bolsaterminal.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class OHLCVBar(
    val time: Double,
    val open: Double,
    val high: Double,
    val low: Double,
    val close: Double,
    val volume: Double,
)

/**
 * Warmup periods (e.g. EMA200 needs 200 bars) produce NaN on the backend,
 * which `JSON.stringify` serializes as `null` — these must stay nullable or
 * decoding throws the moment a shorter period is requested. See the same
 * fix applied to the macOS Swift models (Core/Models/History.swift).
 */
@Serializable
data class IndicatorSet(
    val rsi: List<Double?>,
    val macdLine: List<Double?>,
    val macdSignal: List<Double?>,
    val macdHistogram: List<Double?>,
    val sma20: List<Double?>,
    val sma50: List<Double?>,
    val ema20: List<Double?>,
    val ema200: List<Double?>,
    val bbUpper: List<Double?>,
    val bbMiddle: List<Double?>,
    val bbLower: List<Double?>,
    val bbPercent: List<Double?>,
    val volumeMA20: List<Double?>,
)

@Serializable
enum class SignalType {
    @SerialName("STRONG_BUY") StrongBuy,
    @SerialName("BUY") Buy,
    @SerialName("NEUTRAL") Neutral,
    @SerialName("SELL") Sell,
    @SerialName("STRONG_SELL") StrongSell,
}

@Serializable
data class SignalComponent(
    val name: String,
    val value: Double? = null,
    val signal: String,
    val score: Double,
)

@Serializable
data class Recommendation(
    val signal: SignalType,
    val score: Double,
    val components: List<SignalComponent>,
    val currentRSI: Double? = null,
    val currentMACD: Double? = null,
    val currentBBPercent: Double? = null,
    val priceVsEMA200: Double? = null,
)

@Serializable
data class HistoryResponse(
    val symbol: String,
    val bars: List<OHLCVBar>,
    val indicators: IndicatorSet,
    val recommendation: Recommendation,
)
